compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -2322,18 +2323,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             .iter()
             .map(|e| (e.obligation.predicate, None, Some(e.obligation.cause.clone())))
             .collect();
-        self.suggest_derive(err, &preds);
+        if suggest_derive {
+            self.suggest_derive(err, &preds);
+        } else {
+            // The predicate comes from a binop where the lhs and rhs have different types.
+            let _ = self.note_predicate_source_and_get_derives(err, &preds);
+        }
     }
 
-    pub fn suggest_derive(
+    fn note_predicate_source_and_get_derives(
         &self,
         err: &mut Diagnostic,
         unsatisfied_predicates: &[(
             ty::Predicate<'tcx>,
             Option<ty::Predicate<'tcx>>,
             Option<ObligationCause<'tcx>>,
         )],
-    ) {
+    ) -> Vec<(String, Span, String)> {
```

Comment:
Hmm, it's only in the error path but I'd love to skipp the entire `if let Some(diagnostic_name) = self.tcx.get_diagnostic_name(trait_pred.def_id())` check if `!suggest_derive` but then we'd need to introduce `suggested_derive: bool` to `note_pre_src_and_get_derives` :(